ble_app.h

```diff
@@ -0,0 +1,427 @@
+        return _target_name;
+    }
+
+protected:
+    /**
+     * Sets up adverting payload and start advertising.
+     * This function is invoked when the ble interface is initialized.
+     */
+    void on_init_complete(BLE::InitializationCompleteCallbackContext *event)
+    {
+        if (event->error) {
+            print_error(event->error, "Error during the initialisation\r\n");
+            return;
+        }
+
+        printf("Ble instance initialized\r\n");
+
+        _event_queue.call([this]() { _post_init_cb(_ble, _event_queue); });
+
+        /* All calls are serialised on the user thread through the event queue */
+        _event_queue.call([this]() { start_activity(); });
+    }
+
+    /**
+     * Start the gatt client process when a connection event is received.
+     * This is called by Gap to notify the application we connected
+     */
+    void onConnectionComplete(const ble::ConnectionCompleteEvent &event) override
+    {
+        _is_connecting = false;
+        if (event.getStatus() == BLE_ERROR_NONE) {
+            _connected = true;
+            _conn_handle = event.getConnectionHandle();
+            printf("Connected to: ");
+            print_address(event.getPeerAddress());
+        } else {
+            printf("Failed to connect\r\n");
+            _event_queue.call([this]() { start_activity(); });
+        }
+    }
+
+    /**
+     * Stop the gatt client process when the device is disconnected then restart advertising.
+     * This is called by Gap to notify the application we disconnected
+     */
+    void onDisconnectionComplete(const ble::DisconnectionCompleteEvent &event) override
+    {
+        if (_connected) {
+            _connected = false;
+            printf("Disconnected.\r\n");
+            _event_queue.call([this]() { start_activity(); });
+        }
+    }
+
+    /** Restarts main activity */
+    void onAdvertisingEnd(const ble::AdvertisingEndEvent &event)
+    {
+        _event_queue.call([this]() { start_activity(); });
+    }
+
+    /**
+     * Start advertising or scanning. Triggered by init or disconnection.
+     */
+    virtual void start_activity()
+    {
+        if (!_ble.hasInitialized()) {
+            return;
+        }
+
+        if (_advertising_name) {
+            start_advertising();
+        } else {
+            _ble.gap().stopAdvertising(_adv_handle);
+        }
+
+        if (_target_name) {
+            start_scanning();
+        } else {
+            _ble.gap().stopScan();
+        }
+    }
+
+    /**
+     * Start the advertising process; it ends when a device connects.
+     */
+    void start_advertising()
+    {
+        ble_error_t error;
+
+        if (!_advertising_name || _ble.gap().isAdvertisingActive(_adv_handle)) {
+            /* we're already advertising */
+            return;
+        }
+
+        ble::AdvertisingParameters adv_params(
+            ble::advertising_type_t::CONNECTABLE_UNDIRECTED,
+            ble::adv_interval_t(ble::millisecond_t(40))
```

Comment:
Maybe you should provide a setter for this?

Reply:
Yeah, this is not really production code. Those kind of features will be handled by the real BLE App that we'll be developing soon